### PR TITLE
attempt at fixing become_leader to be more sane

### PR DESCRIPTION
**Pull request by sjpotter**
1) change become_leader to not directly append_entry or send_appendentries() itself, but to rely on recv_entry

this requires restructuring it.

2) this exposed issues in test code where the noop that has it become leader wasn't being comitted, so the count of entries for itself and other nodes was off, had to modify this in multiple places

3) the next_idx == last_log_idx + 1 test is no longer accurate for similar reason, as now the noop is the last_log_idx so it should be next_idx for all nodes

### CODE REVIEW

File: src/raft_server.c [hunk 1]
```diff
@@ -192,9 +179,20 @@ int raft_become_leader(raft_server_t* me_)
...
-        raft_send_appendentries(me_, node);
     }
 
+    raft_entry_t *noop = raft_entry_new(0);
+    noop->term = raft_get_current_term(me_);
+    noop->type = RAFT_LOGTYPE_NO_OP;
+    int e = raft_recv_entry(me_, noop, &response);
+    if (0 != e)
+        return e;
+    raft_entry_release(noop);
```
Comment (tezc) [inline src/raft_server.c:190]:
```suggestion

    int e = raft_recv_entry(me_, noop, &response);
    raft_entry_release(noop);
    
    if (0 != e)
        return e;
```

Let's fix the memory leak in the way.

File: tests/test_server.c [hunk 2]
```diff
@@ -3904,7 +3904,7 @@ int main(void)
     SUITE_ADD_TEST(suite, TestRaft_candidate_recv_appendentries_from_same_term_results_in_step_down);
     SUITE_ADD_TEST(suite, TestRaft_leader_becomes_leader_is_leader);
     SUITE_ADD_TEST(suite, TestRaft_leader_becomes_leader_does_not_clear_voted_for);
-    SUITE_ADD_TEST(suite, TestRaft_leader_when_becomes_leader_all_nodes_have_nextidx_equal_to_lastlog_idx_plus_1);
+    SUITE_ADD_TEST(suite, TestRaft_leader_when_becomes_leader_all_nodes_have_nextidx_equal_to_lastlog_idx);
```
Comment (tezc) [inline tests/test_server.c:3907]:
```c
void single_node_commits_noop_cb(void* arg, int can_read)
{
    (void) can_read;

    *((char**) arg) = "success";
}

void TestRaft_single_node_commits_noop(CuTest * tc)
{
    static char* str = "test";
    void *r = raft_new();

    raft_add_node(r, NULL, 1, 1);
    raft_set_current_term(r, 2);
    raft_set_commit_idx(r, 0);
    raft_periodic(r, 500);
    raft_queue_read_request(r, single_node_commits_noop_cb, &str);
    raft_periodic(r, 500);

    CuAssertIntEquals(tc, 1, raft_get_commit_idx(r));
    CuAssertStrEquals(tc, "success", str);

    raft_free(r);
}
```

We can add this to test original issue.

File: src/raft_server.c (hunk 1 above)
Comment (tezc) [reply src/raft_server.c:190]: Ahh, also we may want to revert back to follower if we get an error here? (I know we are not doing good with error handling in the library right now but let's match the previous behavior, it was setting itself "leader" after the append. If we go back to follower on error, we'll match the previous behavior I believe)

**Comment by tezc** [comment]
Just minor comments, looks good to me.

**Comment by sjpotter** [comment]
we decided against this I believe?

**Pull request closed without merging**